### ISSUE TITLE
Remove Build Timestamp From Shader Asset Data 

### ISSUE THREAD

**Issue by santorac** [o3de/o3de#12959]
**Is your feature request related to a problem? Please describe.**
ShaderAsset data currently includes the timestamp of when the asset was compiled by the Asset Processor. This has been used to prevent edge cases when hot-loading shader variants. It ensures that older shader variants are not used for a more recently updated shader asset. For example, see https://github.com/o3de/o3de/blob/097844658da045ea38ffae7c2ced0c48ca308f19/Gems/Atom/RPI/Code/Source/RPI.Reflect/Shader/ShaderAsset.cpp#L267

The problem is that the ShaderAsset file content is not consistent and deterministic. The binary content of the file will be slightly different each time the shader is compiled. This could cause problems with dependency tracking or packaging systems that would rely on consistent file content within the cache.

**Describe the solution you'd like**
We need to remove m_buildTimestamp from ShaderAsset and ShaderVariantAsset. Any subsequent shader hot-reload issues must be addressed in some other way that avoids any non-determinism in the final shader asset file in the asset cache.

These are the possible issues I'm aware of that would need to be addressed, all related to what happens when hot-loading updated shader variant assets at runtime. I did some testing by locally removing m_buildTimestamp from ShaderAsset and ShaderVariantAsset and this is what I found:

1. **When you change shader code, does the rendered result automatically update?** I actually don't expect any problems here since the Shader class handles hot reload for both ShaderAsset and ShaderVariantAsset. I didn't notice any problems in this area during testing, but it should be tested more thoroughly when committing these changes.

1. **When the rendered result automatically updates, does it take a long time to get there?** The ShaderAsset and ShaderVariantAsset are built by separate Asset Processor jobs. The ShaderAsset builder has a somewhat high priority, while the ShaderVariantAsset has a low priority because there can be many of them and they are merely an optimization. While the ShaderAsset would be processed quickly and updated in memory, the runtime will continue using the Shader**Variant**Asset until that is updated. If the ShaderVariantAsset processing is delayed because many other assets are being processed first (this is common with shader variant assets, because there can be so many of them) then the runtime will continue rendering using the old ShaderVariantAsset which has old shader code, even though a more recent ShaderAsset is available. Consequently, when I tested locally, there were timing failures in AtomSampleViewer's MaterialHotReloadTest, which I had to resolve by adding more time in the "IdleForReload" function. In a real-world scenario, it could take minutes for shader changes to show up instead of seconds. There are a couple ideas for addressing this concern, described below.

1. **When you remove a shader variant from the list, and it disappears from the asset catalog, does it linger in memory?** It isn't typical for runtime code to listen for messages about assets being _removed_, they only care if the assets are _updated_ and show the latest content. Once they are available and loaded into memory, the runtime just continues using them. They won't disappear unless you restart the Editor. This will be especially confusing in the case of shaders and shader variants. If the runtime is using a shader variant to render, and that ShaderVariantAsset is removed, it will continue rendering that shader variant since it's still in memory. Then suppose the user updates some shader code. This will update the ShaderAsset and reload it, but it will not update the ShaderVariantAsset because it was removed. The runtime will continue using the old shader variant even though a newer shader is available. This could also impact performance testing as it won't be possible to repeatedly add and remove a shader variant for comparison. Only after restarting the Editor will the user finally see the update shader behavior. This situation can be seen in AtomSampleViewer's MaterialHotReloadTest by simply running the test multiple times. The first time it will pass, but subsequent times it will fail because it's using shader variants from the prior scenario. The generaly solution here is to make the Shader class listen to AssetCatalogEventBus::OnCatalogAssetRemoved, in addition to OnAssetReloaded.

Now regarding item number 2 above, I see two general approaches we could take:
1. Use a checksum instead of a timestamp. This would be deterministic and can still ensure that ShaderVariantAssets are only used together with the corresponding version of the ShaderAsset. The ShaderAssetBuilder would have to checksum the entire content of the ShaderAsset file, and then insert the checksum into the file before saving. The ShaderVariantAssetBuilder's will need to load the corresponding ShaderAsset, retrieve the checksum, and store the checksum in the ShaderVariantAsset. Then at runtime it would just check the checksums where before it was checking timestamps. Note that the ShaderVariantAssetBuilder's job will need a job depdendency on the ShaderAssetBuilder's job, to make sure it loads the updated version of the asset. (At some point in the past we had a job depenendency like this, and we ended up removing it because of some problem, and I can't remember what that problem was. So just be aware this needs a closer look).
1. Implement some improved prioritization system in the Asset Processor to ensure that the user doesn't have to wait a long time for the shader variant to update. The asset system should know what content the user is viewing, which shader assets and shader variant assets are being used, and prioritize those shader variant jobs above all other shader variant jobs (or don't even process the jobs that aren't needed at the moment). There has been some talk in sig-core about pursuing this kind of usage-based asset processing, but I'm not sure when it will be ready.

**Describe alternatives you've considered**

**Additional context**

**Comment by galibzon** [comment]
This is done
